### DISCUSSION TITLE
Workaround to support multiple db search path (DB_SEARCH_PATH)

### DISCUSSION THREAD

**Discussion by nya1** [directus/directus#12057] [Archive]
Hello,
Just wanted to share my workaround to support multiple search paths (postgresql)

## Background:
I was trying to run directus on a custom schema but I was getting an error (see below) when I was trying to load some tables with geometry type (postgis), I discovered that since I set `DB_SEARCH_PATH=customschema` directus was unable to use postgis extension that was installed in `public` schema, this is the schema setup:

- `public` (postgis extension installed by default)
- `customschema` (where all my normal and directus tables are located)

error:

```jsonc
{
  "errors": [
    {
      "message": "select st_astext(\"MYTABLE\".\"geom\") as \"geom\", \"MYTABLE\".\"id\" from \"MYTABLE\" order by \"MYTABLE\".\"id\" asc limit $1 - function st_astext(public.geometry) does not exist",
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ]
}
```

## Workaround

Since using `array:` syntax in `DB_SEARCH_PATH` env variable will cause a throw on directus startup [1][2], I added a new env variable `DB_CUSTOM_SCHEMA` that allows to set a custom schema for schema introspection and in the same time allows to support `array:` syntax in `DB_SEARCH_PATH`,
in this way directus will use your `DB_CUSTOM_SCHEMA` (where all your tables are located) only for schema introspection but in the same time it's able to access other schemas listed in the `DB_SEARCH_PATH`  (postgis functions in my case)

In `getDatabaseSchema` function (node_modules/directus/dist/utils/get-schema.js) add the following line as the first line of the body function

 - `schemaInspector.explodedSchema = env_1.default.DB_CUSTOM_SCHEMA ? [ env_1.default.DB_CUSTOM_SCHEMA ] : schemaInspector.explodedSchema`

Then in your env configuration you can add the following variables:

```
# add multiple search paths
DB_SEARCH_PATH="array:customschema,public"

# schema where your directus and normal tables are located
DB_CUSTOM_SCHEMA="customschema"
```

For a more "cleaner" patch I've used [patch-package](https://github.com/ds300/patch-package) standard to apply the patch locally, instructions with patch file is located here: https://gist.github.com/nya1/d8f84e9e2999bd2d169ef228daed6cbc

(tested on directus 9.5.1 and 9.6.0)

Related issues
- [1] https://github.com/directus/directus/issues/5397
- [2] https://github.com/directus/directus/discussions/3164#discussioncomment-1600092
- https://github.com/directus/directus/discussions/5395#discussioncomment-680087

**Comment by rijkvanzanten** [comment]
Awesome! Great little workaround to support having the Directus tables in a different schema until we support that natively 🙂

**Comment by productdevbook** [comment]
Thank you working ! 

```ts
      DB_CLIENT: 'pg'
      DB_HOST: 'db'
      DB_PORT: '5432'
      DB_DATABASE: 'postgres'
      DB_SCHEMA: 'directus'
      DB_CUSTOM_SCHEMA: 'directus'
      DB_USER: '${POSTGRES_USER}'
      DB_PASSWORD: '${POSTGRES_PASSWORD}'
      DB_SEARCH_PATH: "array:directus,auth,public"
```
